internal/pipeline/fn.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pipeline
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+
+	"go.uber.org/zap"
+	"sigs.k8s.io/kustomize/kyaml/comments"
+	"sigs.k8s.io/kustomize/kyaml/kio"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+const (
+	// annotation used to track resources for preserving comments during the function execution
+	idAnnotation = "config.k8s.io/id"
+)
+
+// KRMFn defines capabilities of a KRM function. It primarily knows
+// how to run itself. It accepts input from an io.Reader that is
+// a stream of KRM objects in the form of KRM FunctionSpec. It also accepts
+// a io.Rriter where it writes the result of running a KRM function.
+type KRMFn interface {
```

Comment:
General question: Are you extracting a common library for use by 'kpt fn' and 'kpt pipeline' in the kpt repo?